### PR TITLE
Document the RemoveUnusedDeclarations pass.

### PR DESCRIPTION
**Pull request by cole-barefoot**

### CODE REVIEW

File: frontends/p4/unusedDeclarations.h [hunk 1]
```diff
@@ -70,14 +94,22 @@ class RemoveUnusedDeclarations : public Transform {
...
+ * If @warn is true, emit compiler warnings if an unused instance of an
+ * IR::P4Table or IR::Declaration_Instance is removed.
+ */
 class RemoveAllUnusedDeclarations : public PassManager {
  public:
     explicit RemoveAllUnusedDeclarations(ReferenceMap* refMap, bool warn = false) {
         CHECK_NULL(refMap);
         std::set<const IR::Node*> *warned = nullptr;
+
+        // TODO: The @warned set is discarded.  Is it necessary, or can it be replaced
```
Comment (mbudiu-vmw) [inline frontends/p4/unusedDeclarations.h]: The set is not discarded, it is used as an argument for the RemoveUnusedDeclarations below.
It is the same set for all iterations of the PassRepeated, thus it cannot be replaced with a bool.

Comment (cole-barefoot) [inline frontends/p4/unusedDeclarations.h]: Oh I see.  This avoids issuing multiple warnings for the same unreferenced extern definition, which is the only thing that can both trigger a warning and persist across multiple rounds.

I'll remove the TODO and add a comment to this effect.  Thanks for pointing it out.

File: frontends/p4/unusedDeclarations.h [hunk 2]
```diff
@@ -22,10 +22,38 @@ limitations under the License.
...
+ * The following kinds of nodes are not removed even if they are unreferenced:
+ *  - IR::Declaration_MatchKind
+ *  - IR::Parameter
+ *  - IR::Type_Error
+ *  - IR::Type_Extern
+ *  - IR::Type_Method
+ *  - IR::Type_StructLike
+ *
+ * If @warned is non-null, unused IR::P4Table and IR::Declaration_Instance
+ * nodes are stored in @warned if they are unused and removed by this pass.
```
Comment (ChrisDodd) [inline frontends/p4/unusedDeclarations.h]: In addition, a warning message is emitted only if the declaration was not previously in `warned` -- that is in fact the main point of `warned` -- to suppress duplicate warnings

**Comment by cole-barefoot** [comment]
Thanks @mbudiu-vmw and @ChrisDodd.  I updated the comments to (a) remove the TODO, which was wrong, and (b) clarify the behavior of the warning list.

**Pull request merged**